server/bin/pbench-cull-unpacked-tarballs.py

```diff
@@ -0,0 +1,457 @@
+        )
+        return 4
+
+    incoming = config.INCOMING
+    incoming_p = os.path.realpath(incoming)
+
+    if not incoming_p:
+        logger.error("The configured INCOMING directory, {}, does not exist", incoming)
+        return 3
+
+    if not os.path.isdir(incoming_p):
+        logger.error(
+            "The configured INCOMING directory, {}, is not a valid directory", incoming
+        )
+        return 4
+
+    results = config.RESULTS
+    results_p = os.path.realpath(results)
+
+    if not results_p:
+        logger.error("The configured RESULTS directory, {}, does not exist", results)
+        return 3
+
+    if not os.path.isdir(results_p):
+        logger.error(
+            "The configured RESULTS directory, {}, is not a valid directory", results
+        )
+        return 4
+
+    users = config.USERS
+    users_p = os.path.realpath(users)
+
+    if not users_p:
+        logger.error("The configured USERS directory, {}, does not exist", users)
+        return 3
+
+    if not os.path.isdir(users_p):
+        logger.error(
+            "The configured USERS directory, {}, is not a valid directory", users
+        )
+        return 4
+
+    # Fetch the configured maximum number of days a tar can remain "unpacked"
+    # in the INCOMING tree.
+    try:
+        max_unpacked_age = config.conf.get("pbench-server", "max-unpacked-age")
+    except NoOptionError as e:
+        logger.error(f"{e}")
+        return 5
+    try:
+        max_unpacked_age = int(max_unpacked_age)
+    except Exception:
+        logger.error("Bad maximum unpacked age, {}", max_unpacked_age)
+        return 6
+
+    # First phase is to find all the tar balls which are beyond the max
+    # unpacked age, and which still have an unpacked directory in INCOMING.
+    if config._ref_datetime is not None:
+        try:
+            curr_dt = config._ref_datetime
+        except Exception:
+            # Ignore bad dates from test environment.
+            curr_dt = datetime.utcnow()
+    else:
+        curr_dt = datetime.utcnow()
+
+    _msg = "Culling unpacked tar balls {} days older than {}"
+    if options.dry_run:
+        print(_msg.format(max_unpacked_age, curr_dt.strftime(_STD_DATETIME_FMT)))
+    else:
+        logger.debug(_msg, max_unpacked_age, curr_dt.strftime(_STD_DATETIME_FMT))
+
+    actions_taken = []
+    errors = 0
+    start = pbench._time()
+    for tb_incoming_dir, controller_name in gen_list_unpacked_aged(
+        incoming_p, archive_p, curr_dt, max_unpacked_age
+    ):
+        errors, _actions_taken, duration = remove_unpacked(
+            tb_incoming_dir,
+            controller_name,
+            results_p,
+            users_p,
+            logger,
+            options.dry_run,
+        )
+        unpacked_dir_name = os.path.basename(tb_incoming_dir)
+        act_name = os.path.join(controller_name, unpacked_dir_name)
+        actions_taken.append((_actions_taken, act_name, errors, duration))
+        if errors > 0:
```

Comment:
So if one unpacked tar ball fails to be removed, then we abort the whole thing? Why not try them all?

Reply:
Because we are deleting data which could be catastrophic if the errors are due to a bad configuration.  This is very conservative for sure, but seems like the best approach until we get enough experience with the errors we encounter.  We can always adjust this later when we gain familiarity with how this new service will run.

Reply:
OK